File: engine/src/main/java/de/hanno/hpengine/engine/graphics/shader/ShaderDefine.java
package de.hanno.hpengine.engine.graphics.shader;

import de.hanno.hpengine.engine.config.Config;
import de.hanno.hpengine.engine.model.material.Material.MAP;

import java.util.EnumSet;

public enum ShaderDefine {
	
	DIFFUSE(MAP.DIFFUSE),
	NORMAL(MAP.NORMAL),
	SPECULAR(MAP.SPECULAR),
	OCCLUSION(MAP.DISPLACEMENT),
	HEIGHT(MAP.HEIGHT),
	REFLECTION(MAP.REFLECTION),
	ROUGHNESS(MAP.ROUGHNESS);

	private String defineString;
	private MAP map;
	
	public String getDefineText() {
		return defineString;
	};
	
	ShaderDefine(MAP map) {
		this.map = map;
		this.defineString = "#define use_" + map.getShaderVariableName();
	}
	
	public static String getDefineString(EnumSet<ShaderDefine> defines) {
		if (defines == null) {
			return "";
		}
		StringBuilder builder = new StringBuilder();
		
		for (ShaderDefine shaderDefine : defines) {
			builder.append(shaderDefine.getDefineText());
			builder.append("\n");
		}
		return builder.toString();
	}

	public static String getGlobalDefinesString(Config config) {
		StringBuilder builder = new StringBuilder();

		appendWithSemicolonAndNewLine(builder, "const bool RAINEFFECT = " + (config.getEffects().getRainEffect() != 0.0));
		appendWithSemicolonAndNewLine(builder, "const bool MULTIPLE_DIFFUSE_SAMPLES = " + config.getQuality().isUseMultipleDiffuseSamples());
		appendWithSemicolonAndNewLine(builder, "const bool MULTIPLE_DIFFUSE_SAMPLES_PROBES = " + config.getQuality().isUseMultipleDiffuseSamplesProbes());
		appendWithSemicolonAndNewLine(builder, "const bool USE_CONETRACING_FOR_DIFFUSE = " + config.getQuality().isUseConetracingForDiffuse());
		appendWithSemicolonAndNewLine(builder, "const bool USE_CONETRACING_FOR_DIFFUSE_PROBES = " + config.getQuality().isUseConetracingForDiffuseProbes());
		appendWithSemicolonAndNewLine(builder, "const bool USE_CONETRACING_FOR_SPECULAR = " + config.getQuality().isUseConetracingForSpecular());
		appendWithSemicolonAndNewLine(builder, "const bool USE_CONETRACING_FOR_SPECULAR_PROBES = " + config.getQuality().isUseConetracingForSpecularProbes());
		appendWithSemicolonAndNewLine(builder, "const bool PRECOMPUTED_RADIANCE = " + config.getQuality().isUsePrecomputedRadiance());
		appendWithSemicolonAndNewLine(builder, "const bool SCATTERING = " + config.getEffects().isScattering());
		appendWithSemicolonAndNewLine(builder, "const bool CALCULATE_ACTUAL_RADIANCE = " + config.getQuality().isCalculateActualRadiance());
		appendWithSemicolonAndNewLine(builder, "const bool SSR_FADE_TO_SCREEN_BORDERS = " + config.getQuality().isSsrFadeToScreenBorders());
		appendWithSemicolonAndNewLine(builder, "const bool SSR_TEMPORAL_FILTERING = " + config.getQuality().isSsrTemporalFiltering());
		appendWithSemicolonAndNewLine(builder, "const bool USE_BLOOM = " + config.getEffects().isUseBloom());
		appendWithSemicolonAndNewLine(builder, "const bool USE_PCF = " + config.getQuality().isUsePcf());
		appendWithSemicolonAndNewLine(builder, "const bool USE_DPSM = " + config.getQuality().isUseDpsm());
		builder.append("\n");
		return builder.toString();
	}
	
	private static StringBuilder appendWithSemicolonAndNewLine(StringBuilder builder, String toAppend) {
		builder.append(toAppend);
		builder.append(";\n");
		return builder;
	}
}

File: engine/src/main/kotlin/de/hanno/hpengine/engine/graphics/shader/ShaderDefine.kt
package de.hanno.hpengine.engine.graphics.shader

import de.hanno.hpengine.engine.config.Config
import de.hanno.hpengine.engine.model.material.Material.MAP
import java.util.EnumSet
import java.lang.StringBuilder

enum class ShaderDefine(private val map: MAP) {
    DIFFUSE(MAP.DIFFUSE), NORMAL(MAP.NORMAL), SPECULAR(MAP.SPECULAR), OCCLUSION(MAP.DISPLACEMENT), HEIGHT(MAP.HEIGHT), REFLECTION(
        MAP.REFLECTION
    ),
    ROUGHNESS(MAP.ROUGHNESS);

    val defineText: String = "#define use_" + map.shaderVariableName

    companion object {
        fun getDefineString(defines: EnumSet<ShaderDefine>?): String {
            if (defines == null) {
                return ""
            }
            val builder = StringBuilder()
            for (shaderDefine in defines) {
                builder.append(shaderDefine.defineText)
                builder.append("\n")
            }
            return builder.toString()
        }

        fun getGlobalDefinesString(config: Config): String {
            val builder = StringBuilder()
            appendWithSemicolonAndNewLine(
                builder,
                "const bool RAINEFFECT = " + (config.effects.rainEffect.toDouble() != 0.0)
            )
            appendWithSemicolonAndNewLine(
                builder,
                "const bool MULTIPLE_DIFFUSE_SAMPLES = " + config.quality.isUseMultipleDiffuseSamples
            )
            appendWithSemicolonAndNewLine(
                builder,
                "const bool MULTIPLE_DIFFUSE_SAMPLES_PROBES = " + config.quality.isUseMultipleDiffuseSamplesProbes
            )
            appendWithSemicolonAndNewLine(
                builder,
                "const bool USE_CONETRACING_FOR_DIFFUSE = " + config.quality.isUseConetracingForDiffuse
            )
            appendWithSemicolonAndNewLine(
                builder,
                "const bool USE_CONETRACING_FOR_DIFFUSE_PROBES = " + config.quality.isUseConetracingForDiffuseProbes
            )
            appendWithSemicolonAndNewLine(
                builder,
                "const bool USE_CONETRACING_FOR_SPECULAR = " + config.quality.isUseConetracingForSpecular
            )
            appendWithSemicolonAndNewLine(
                builder,
                "const bool USE_CONETRACING_FOR_SPECULAR_PROBES = " + config.quality.isUseConetracingForSpecularProbes
            )
            appendWithSemicolonAndNewLine(
                builder,
                "const bool PRECOMPUTED_RADIANCE = " + config.quality.isUsePrecomputedRadiance
            )
            appendWithSemicolonAndNewLine(builder, "const bool SCATTERING = " + config.effects.isScattering)
            appendWithSemicolonAndNewLine(
                builder,
                "const bool CALCULATE_ACTUAL_RADIANCE = " + config.quality.isCalculateActualRadiance
            )
            appendWithSemicolonAndNewLine(
                builder,
                "const bool SSR_FADE_TO_SCREEN_BORDERS = " + config.quality.isSsrFadeToScreenBorders
            )
            appendWithSemicolonAndNewLine(
                builder,
                "const bool SSR_TEMPORAL_FILTERING = " + config.quality.isSsrTemporalFiltering
            )
            appendWithSemicolonAndNewLine(builder, "const bool USE_BLOOM = " + config.effects.isUseBloom)
            appendWithSemicolonAndNewLine(builder, "const bool USE_PCF = " + config.quality.isUsePcf)
            appendWithSemicolonAndNewLine(builder, "const bool USE_DPSM = " + config.quality.isUseDpsm)
            builder.append("\n")
            return builder.toString()
        }

        private fun appendWithSemicolonAndNewLine(builder: StringBuilder, toAppend: String): StringBuilder {
            builder.append(toAppend)
            builder.append(";\n")
            return builder
        }
    }

}